src/appleseed/renderer/modeling/bsdf/glossylayerbsdf.cpp

```diff
@@ -0,0 +1,513 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "glossylayerbsdf.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/lighting/scatteringmode.h"
+#include "renderer/kernel/shading/directshadingcomponents.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/modeling/bsdf/bsdf.h"
+#include "renderer/modeling/bsdf/bsdfsample.h"
+#include "renderer/modeling/bsdf/bsdfwrapper.h"
+#include "renderer/modeling/bsdf/energycompensation.h"
+#include "renderer/modeling/bsdf/energycompensationtables.h"
+#include "renderer/modeling/bsdf/fresnel.h"
+#include "renderer/modeling/bsdf/microfacethelper.h"
+#include "renderer/utility/messagecontext.h"
+#include "renderer/utility/paramarray.h"
+
+// appleseed.foundation headers.
+#include "foundation/containers/dictionary.h"
+#include "foundation/math/basis.h"
+#include "foundation/math/microfacet.h"
+#include "foundation/math/minmax.h"
+#include "foundation/math/sampling/mappings.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/makevector.h"
+#include "foundation/utility/otherwise.h"
+
+// Boost headers.
+#include "boost/filesystem/path.hpp"
+
+// Standard headers.
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <string>
+
+// Forward declarations.
+namespace foundation    { class IAbortSwitch; }
+namespace renderer      { class Assembly; }
+namespace renderer      { class Project; }
+
+using namespace foundation;
+using namespace std;
+
+namespace bfs = boost::filesystem;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Glossy Layer BSDF.
+    //
+    // References:
+    //
+    //   [1] Microfacet Models for Refraction through Rough Surfaces
+    //       http://www.cs.cornell.edu/~srm/publications/EGSR07-btdf.pdf
+    //
+    //   [2] Physically-Based Shading at Disney
+    //       https://disney-animation.s3.amazonaws.com/library/s2012_pbs_disney_brdf_notes_v2.pdf
+    //
+    //   [3] Revisiting Physically Based Shading at Imageworks
+    //       http://blog.selfshadow.com/publications/s2017-shading-course/imageworks/s2017_pbs_imageworks_slides.pdf
+    //
+
+    const char* Model = "glossy_layer_bsdf";
+
+    class GlossyLayerBSDFImpl
+      : public BSDF
+    {
+      public:
+        GlossyLayerBSDFImpl(
+            const char*                 name,
+            const ParamArray&           params)
+          : BSDF(name, Reflective, ScatteringMode::Glossy, params)
+        {
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override
+        {
+            return Model;
+        }
+
+        size_t compute_input_data_size() const override
+        {
+            return sizeof(InputValues);
+        }
+
+        void prepare_inputs(
```

Comment:
The Glossy BRDF does this:

    values->m_roughness = std::max(values->m_roughness, shading_point.get_ray().m_min_roughness);

Isn't it necessary here too for roughness clamping to work?